### PR TITLE
`resultTransformer` and `resultComparator`

### PR DESCRIPTION
**Pull request by nevir**
You can provide a `resultTransformer` when configuring `ApolloClient` to be able to modify query/mutation results immediately before they are returned to the application.  Generally, transformers should avoid mutations of their input, as it will bust Apollo's result equality comparison (and cause unnecessary re-renders).

For cases where it is preferable to mutate the input (performance), you can also override how query results are compared for equality via `resultComparator`.  This compares previously returned query results with newly fetched ones.

As an example, an application may wish to attach prototypes to result data (for read-only helpers) - it would need to modify the equality check to deep compare the two results while ignoring prototypes.  If the newly returned result (without prototypes) matches the data present in the previous result (with prototypes), Apollo would consider it a no-op and not trigger any changes.

---

This is a continuation of #378

Also, note that the prototype example is how we are currently using this transformer in our application

### CODE REVIEW

File: src/QueryManager.ts [hunk 1]
```diff
@@ -249,7 +273,10 @@ export class QueryManager {
           ],
         });
 
-        return result;
+        // data is optional in GraphQLResult, but not ApolloQueryResult.
+        const apolloResult = {data: result.data};
```
Comment (nevir) [inline src/QueryManager.ts]: It's not clear to me why TypeScript missed this previously (`mutate` has a return type of `Promise<ApolloQueryResult>`, and was previously resolving this promise with a `GraphQLResult`)

Comment (Poincare) [inline src/QueryManager.ts]: That's probably because `GraphQLResult` is the same thing as an `ApolloQueryResult` but with an extra field tacked on.

Comment (nevir) [inline src/QueryManager.ts]: The `data` property is optional in `GraphQLResult`, but required in `ApolloQueryResult`

Comment (nevir) [inline src/QueryManager.ts]: Lemme bump this to just cast the value rather than constructing a new object

**Comment by stubailo** [comment]
> For cases where it is preferable to mutate the input

Is there ever a good reason to mutate it? I guess just to avoid cloning overhead?

**Comment by nevir** [comment]
> Is there ever a good reason to mutate it? I guess just to avoid cloning overhead?

Yeah, I really worry about cloning the entire graph every time we pull down some values (again; since it's already being "cloned" as the values are being denormalized from the store)

**Comment by nevir** [comment]
Hmm, AFAIK @igrayson did sign the CLA

**Comment by igrayson** [comment]
Any feedback on this?

**Comment by stubailo** [comment]
This looks fine to me! Sorry for taking a long time to look at it. If you rebase on top of current master I will merge it.

**Comment by nevir** [comment]
@stubailo rebased!

**Comment by igrayson** [comment]
@stubailo another ping :>

**Comment by igrayson** [comment]
@stubailo ping

**Comment by stubailo** [comment]
Merged! 🎉 

Sorry it took so long.

**Pull request closed without merging**